File: backend/src/main/java/com/hrms/services/LeaveService.java
package com.hrms.services;

import java.util.List;

import com.hrms.dtos.request.LeaveRequestDTO;
import com.hrms.dtos.response.LeaveResponseDTO;
import com.hrms.enums.LeaveStatus;

public interface LeaveService {

    // Apply leave
    LeaveResponseDTO applyLeave(
            LeaveRequestDTO leaveRequestDTO);

    // Get all leave requests
    List<LeaveResponseDTO> getAllLeaves();

    // Employee leave history
    List<LeaveResponseDTO> getLeavesByEmployee(
            Long employeeId);

    // Filter by status
    List<LeaveResponseDTO> getLeavesByStatus(
            LeaveStatus status);

    // Approve leave
    LeaveResponseDTO approveLeave(
            Long leaveId);

    // Reject leave
    LeaveResponseDTO rejectLeave(
            Long leaveId);
    
    List<LeaveResponseDTO> getMyLeaves();
}

File: frontend/src/services/leaveService.js
import api from "./api";

export const leaveService = {

    // ADMIN / HR
    getAllLeaves: async () => {
        const response = await api.get("/leaves");
        return response.data;
    },

    // EMPLOYEE - logged-in employee only
    getMyLeaves: async () => {
        const response = await api.get("/leaves/me");
        return response.data;
    },

    // EMPLOYEE - apply leave
    applyLeave: async (leaveData) => {

        const payload = {
            leaveType: leaveData.leaveType,
            startDate: leaveData.startDate,
            endDate: leaveData.endDate,
            reason: leaveData.reason
        };

        const response = await api.post("/leaves", payload);

        return response.data;
    },

    // ADMIN / HR / MANAGER
    getLeavesByEmployee: async (employeeId) => {
        const response =
            await api.get(`/leaves/employee/${employeeId}`);

        return response.data;
    },

    // ADMIN / HR / MANAGER
    getLeavesByStatus: async (status) => {
        const response =
            await api.get("/leaves/status", {
                params: { status }
            });

        return response.data;
    },

    // ADMIN / HR / MANAGER
    approveLeave: async (leaveId) => {
        const response =
            await api.put(`/leaves/${leaveId}/approve`);

        return response.data;
    },

    // ADMIN / HR / MANAGER
    rejectLeave: async (leaveId) => {
        const response =
            await api.put(`/leaves/${leaveId}/reject`);

        return response.data;
    }
};